src/Language/PureScript/Types.hs

```diff
@@ -598,10 +607,27 @@ quantify ty = foldr (\arg t -> ForAll (getAnnForType ty) TypeVarInvisible arg No
 
 -- | Move all universal quantifiers to the front of a type
 moveQuantifiersToFront :: Type a -> Type a
-moveQuantifiersToFront = go [] [] where
-  go qs cs (ForAll ann vis q mbK ty sco) = go ((ann, q, sco, mbK, vis) : qs) cs ty
-  go qs cs (ConstrainedType ann c ty) = go qs ((ann, c) : cs) ty
-  go qs cs ty = foldl (\ty' (ann, q, sco, mbK, vis) -> ForAll ann vis q mbK ty' sco) (foldl (\ty' (ann, c) -> ConstrainedType ann c ty') ty cs) qs
+moveQuantifiersToFront = go [] [] 
+  where
+  go qs cs = \case
+    ForAll ann vis q mbK ty sco -> do
+      let 
+        cArgs :: [Text] = cs >>= mapMaybe tyVarText . constraintArgs . snd
+        (q'', ty')
+          | q `elem` cArgs = do
+              let q' = genPureName q cArgs
```

Comment:
The second argument to `genPureName` also has to contain any free type variables in `ty`, I think.

Reply:
Given that `ty` is
`forall a. a`
which translates to 
`ForAll ann TyVarInvisible "a" Nothing (TyVar ann2 "a") Nothing`, wouldn't `freeTypeVariables ty` in this case include `a` as a free type variable? In which case, we would always generate a new name for every `ForAll`, right?

Reply:
> wouldn't `freeTypeVariables ty` in this case include `a` as a free type variable?

Yes.

> In which case, we would always generate a new name for every `ForAll`, right?

No, because we're still only doing the renaming if ``q `elem` cArgs``. We just have to make sure that in a context where we do have to generate a new name, such as `forall a. Monoid a => forall a. Tuple a a0`, we don't rename the `a` to `a0` and end up with `forall a a0. Monoid a => Tuple a0 a0`, which is a different type.

Reply:
Ah, now I see. Thanks.